File: 0067-add-binary/0067-add-binary.cpp
class Solution
{
public:
    string addBinary(string a, string b)
    {
        string result = "";
        int p1 = a.size() - 1;
        int p2 = b.size() - 1;
        int carry = 0;
        while (p1 >= 0 || p2 >= 0)
        {
            int bit1 = 0;
            int bit2 = 0;

            if (p1 >= 0)
                bit1 = a[p1--] - '0';

            if (p2 >= 0)
                bit2 = b[p2--] - '0';
            int addition = bit1 + bit2 + carry;
            int res = addition % 2;
            carry = addition / 2;
            result += res + '0';
        }
        result += carry + '0';
        reverse(result.begin(), result.end());
        int firstOne = result.find_first_of('1');
        if (firstOne == -1)
            return "0";
        return result.substr(firstOne);
    }
};

File: 0067-add-binary/0067-add-binary.java
class Solution {
    public String addBinary(String a, String b) {
        int maxLen = Math.max(a.length(), b.length());
        int n = a.length();
        int m = b.length();

        StringBuilder result = new StringBuilder();
        int carry = 0;
        for (int i = 0; i < maxLen; i++) {
            if (i < n) carry += a.charAt(n - 1 - i) - '0';
            if (i < m) carry += b.charAt(m - 1 - i) - '0';
            result.append(carry % 2);
            carry /= 2;
        }
        if (carry != 0)
            result.append(carry);
        return result.reverse().toString();
    }
}